### ISSUE TITLE
mosh won't build for me, protobuf can't find it's input file.

### ISSUE THREAD

**Issue by hartzell** [LLNL/spack#3771]
On a CentOS 7 system with a spack tree from yesterday afternoon, I'm unable to `spack install mosh`.

Jump down to **TL;DR** for an explicit question...

Here's the juicy but of `spack-build.log`:

```
[...]
==> 'make' '-j28'
make  all-recursive
make[1]: Entering directory `/tmp/hartzelg/spack-stage/spack-stage-qPQ1ET/mosh-1.3.0'
Making all in scripts
make[2]: Entering directory `/tmp/hartzelg/spack-stage/spack-stage-qPQ1ET/mosh-1.3.0/scripts'
make[2]: Leaving directory `/tmp/hartzelg/spack-stage/spack-stage-qPQ1ET/mosh-1.3.0/scripts'
Making all in src
make[2]: Entering directory `/tmp/hartzelg/spack-stage/spack-stage-qPQ1ET/mosh-1.3.0/src'
Making all in protobufs
make[3]: Entering directory `/tmp/hartzelg/spack-stage/spack-stage-qPQ1ET/mosh-1.3.0/src/protobufs'
  GEN      userinput.pb.cc
  GEN      hostinput.pb.cc
  GEN      transportinstruction.pb.cc
transportinstruction.protohostinput.proto: File does not reside within any path specified using --proto_path (or -I).  You must specify a --proto_path which encompasses this file.  Note that the proto_path must be an exact prefix of the .proto file names -- protoc is too dumb to figure out when two paths (e.g. absolute and relative) are equivalent (it's harder than you think).: File does not reside within any path specified using --proto_path (or -I).  You must specify a --proto_path which encompasses this file.  Note that the proto_path must be an exact prefix of the .proto file names -- protoc is too dumb to figure out when two paths (e.g. absolute and relative) are equivalent (it's harder than you think).

userinput.proto: File does not reside within any path specified using --proto_path (or -I).  You must specify a --proto_path which encompasses this file.  Note that the proto_path must be an exact prefix of the .proto file names -- protoc is too dumb to figure out when two paths (e.g. absolute and relative) are equivalent (it's harder than you think).
make[3]: *** [transportinstruction.pb.cc] Error 1
[...]
```

I can start a shell w/ the right environment (`spack env mosh bash`) and replicate the failure:

```
[hartzelg@lb046dev protobufs]$ make V=1
/home/hartzelg/tmp/spack-mosh/opt/spack/linux-centos7-x86_64/gcc-4.8.5/protobuf-3.2.0-czzwxzs2ugdrauo4z6lhxvuyqfxz6zsu/bin/protoc --cpp_out=. -I/home/hartzelg/tmp/spack-mosh/var/spack/stage/mosh-1.3.0-fwqhl7dbr67goilqwklz6aqrlxixmakz/mosh-1.3.0/src/protobufs userinput.proto
userinput.proto: File does not reside within any path specified using --proto_path (or -I).  You must specify a --proto_path which encompasses this file.  Note that the proto_path must be an exact prefix of the .proto file names -- protoc is too dumb to figure out when two paths (e.g. absolute and relative) are equivalent (it's harder than you think).
make: *** [userinput.pb.cc] Error 1
```

What's "funny" is that all three of the invocations below work.  It seems that the pathname for the source really has to match, *exactly*, the pathname supplied via `-I` (even to the point of dealing with the fact that the build dir is a symlink).

```
[hartzelg@lb046dev protobufs]$ /home/hartzelg/tmp/spack-mosh/opt/spack/linux-centos7-x86_64/gcc-4.8.5/protobuf-3.2.0-czzwxzs2ugdrauo4z6lhxvuyqfxz6zsu/bin/protoc --cpp_out=. -I/home/hartzelg/tmp/spack-mosh/var/spack/stage/mosh-1.3.0-fwqhl7dbr67goilqwklz6aqrlxixmakz/mosh-1.3.0/src/protobufs /home/hartzelg/tmp/spack-mosh/var/spack/stage/mosh-1.3.0-fwqhl7dbr67goilqwklz6aqrlxixmakz/mosh-1.3.0/src/protobufs/userinput.proto
[hartzelg@lb046dev protobufs]$ /home/hartzelg/tmp/spack-mosh/opt/spack/linux-centos7-x86_64/gcc-4.8.5/protobuf-3.2.0-czzwxzs2ugdrauo4z6lhxvuyqfxz6zsu/bin/protoc --cpp_out=. -I. userinput.proto
[hartzelg@lb046dev protobufs]$ /home/hartzelg/tmp/spack-mosh/opt/spack/linux-centos7-x86_64/gcc-4.8.5/protobuf-3.2.0-czzwxzs2ugdrauo4z6lhxvuyqfxz6zsu/bin/protoc --cpp_out=. -I`pwd` `pwd`/userinput.proto
[hartzelg@lb046dev protobufs]$
```

Here's the bit of the `Makefile.am` that generates the `Makefile.in` that generates the Makefile.  If/when $(srcdir) is anything other than `.`, I don't see how this works.

```
.proto.pb.cc:
    $(AM_V_GEN)$(PROTOC) --cpp_out=. -I$(srcdir) $<
```

If I jump up to the top of the src tree, `make clean`, and then `./configure --prefix=/tmp/moose`, `make` will succeed.  Investigating futher, if I cd into the `src/protobuf` directory and run `make V=1` it ends up invoking `protoc` like so:

```
[hartzelg@lb046dev protobufs]$ make V=1
/home/hartzelg/tmp/spack-mosh/opt/spack/linux-centos7-x86_64/gcc-4.8.5/protobuf-3.2.0-czzwxzs2ugdrauo4z6lhxvuyqfxz6zsu/bin/protoc --cpp_out=. -I. userinput.proto
/home/hartzelg/tmp/spack-mosh/opt/spack/linux-centos7-x86_64/gcc-4.8.5/protobuf-3.2.0-czzwxzs2ugdrauo4z6lhxvuyqfxz6zsu/bin/protoc --cpp_out=. -I. hostinput.proto
/home/hartzelg/tmp/spack-mosh/opt/spack/linux-centos7-x86_64/gcc-4.8.5/protobuf-3.2.0-czzwxzs2ugdrauo4z6lhxvuyqfxz6zsu/bin/protoc --cpp_out=. -I. transportinstruction.proto
make  all-am
```

Notice that now the include path and the source file name are "agreeable".

### TL;DR

It turns out that making this change:

```diff
@@ -43,3 +44,5 @@ class Mosh(AutotoolsPackage):
     depends_on('openssl')

     depends_on('perl', type='run')
+
+    configure_directory='.'
```

to `.../mosh/package.py` allows it to build.

That seems nasty though.  What's the Right Thing To Do?  Does it work on other systems as is?

**Comment by hartzell** [comment]
@adamjstewart -- is this perhaps the result of an AutotoolsPackage update since the mosh package was minted?  Would the change above make sense or is it silliness?

**Comment by adamjstewart** [comment]
Yes, this is likely due to the switch to `AutotoolsPackage`. The way that we usually handle these is to build out-of-source, like:
```python
build_directory = 'spack-build'
```
Your solution would also work. Either are fine to submit a PR for.

**Comment by adamjstewart** [comment]
Closed by #3773.
